openedx/openedx-platform
# feat: Add Library Collections REST endpoints [FC-0062]

## Description

This PR adds REST endpoints (list/create/update/delete) for Library Collections, in addition to emitting events when a library is created/updated. The deleted endpoint and emitting its event will be implemented in a later ticket, however the implementation skeleton is there.

## Supporting information

Related Tickets:
 - https://github.com/openedx/modular-learning/issues/226

Depends on / blocked by:
 - https://github.com/openedx/openedx-events/pull/386
 - https://github.com/openedx/openedx-learning/pull/223

## Testing instructions

1. Go over new tests, make sure they cover most cases and make sure they are passing
1. For additional manual testing:
    1. Checkout this branch in your local dev environment
    1. Make sure you have https://github.com/openedx/openedx-events/pull/386 + https://github.com/openedx/openedx-learning/pull/223 installed and running locally 
    1. Make sure you already have a V2 library or created one if not, e.g `lib:SampleTaxonomyOrg1:AL1`
    1. Add some blocks to that library.
    1. In an external HTTP client (like curl/postman or similar) make the following requests (make sure you have the correct headers for authentication, you can copy them from the network tab in your browser), replace the library key in the URL with your actual library key:
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/ - Should return a list of all collections that belong to the library, initially it will be empty
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/ - Should return a specific collection (with key="COL1") that belongs to the library, initially this should return 404, not found
        1. POST http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/
            1. Payload: {"key": "COL1", "title":"Collection 1", "description": "Desc for Collection 1"} 
            1. This should return the newly created collection
        1. PATCH http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/
            1. Payload: {"title": "Collection 1 updated title"}
            1. This should return the updated collection
        1. POST http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/
            1. Payload: {"key": "COL2", "title":"Collection 2", "description": "Desc for Collection 2"} 
            1. Should return another newly created collection
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/ - This should now return the 2 collections created above
        1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/ - Should return the first collection we created
        1. DELETE http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/ - Should not delete and return a 405
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/blocks/
          1. Should return the list of blocks added to this library
          1. Note their usage key values for the steps below.
       1. PATCH http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/contents/
           1. Payload: `{"usage_keys": [ <list of usage keys from the library blocks> ]`
           1. Should return `{"count": <number of components in list>}`
           1. Check the logs and confirm that the event was emitted for each component, triggering an "upsert" into Meilisearch.
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/
           1. The `modified` timestamp should be updated.
       1. DELETE http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/contents/
           1. Payload: `{"usage_keys": [ <list of usage keys from the library blocks> ]`
           1. Should return `{"count": <number of components in list>}`
           1. Check the logs and confirm that the event was emitted for each component, triggering an "upsert" into Meilisearch.
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/
           1. The `modified` timestamp should be updated.
       1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/COL1/contents/ - will raise a 405, as an object's collections will be fetched from Meilisearch.

---
Private-Ref: [FAL-3783](https://tasks.opencraft.com/browse/FAL-3783)
